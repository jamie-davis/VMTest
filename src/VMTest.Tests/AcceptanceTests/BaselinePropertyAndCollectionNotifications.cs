using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using ApprovalTests;
using ApprovalTests.Reporters;
using NUnit.Framework;
using TestConsoleLib;
using VMTest.Tests.Annotations;
using VMTest.Tests.TestingUtilities;

namespace VMTest.Tests.AcceptanceTests
{
    [TestFixture]
    [UseReporter(typeof (CustomReporter))]
    public class BaselinePropertyAndCollectionNotifications
    {
        private VMMonitor _monitor;

        #region Types for test

        class CollectionMember : AcceptanceTestVM
        {
            private string _name;
            private int _count;

            public string Name
            {
                get { return _name; }
                set
                {
                    if (value == _name) return;
                    _name = value;
                    OnPropertyChanged("Name");
                }
            }

            public int Count
            {
                get { return _count; }
                set
                {
                    if (value == _count) return;
                    _count = value;
                    OnPropertyChanged("Count");
                }
            }
        }

        class MainVM : AcceptanceTestVM
        {
            private ObservableCollection<CollectionMember> _members;

            public ObservableCollection<CollectionMember> Members
            {
                get { return _members; }
                set
                {
                    if (Equals(value, _members)) return;
                    _members = value;
                    OnPropertyChanged("Members");
                }
            }

            public void AddMember(string name, int count)
            {
                Members.Add(new CollectionMember
                {
                    Name = name,
                    Count = count
                });
            }

            public MainVM()
            {
                Members = new ObservableCollection<CollectionMember>();
                var members = new[]
                {
                    "Alpha", "Beta", "Charlie",
                    "Delta", "Echo", "Hotel",
                    "Indigo", "Juliette", "Kilo",
                    "Lima", "Mike", "November",
                    "Oscar", "Papa", "Quebec",
                    "Romeo", "Sierra", "Tango",
                    "Uniform", "Victor", "Whiskey",
                    "X-Ray", "Zulu"
                }
                .Select(n => new CollectionMember
                    {Name = n, Count = n.Length});
                foreach (var member in members)
                {
                    Members.Add(member);
                }
            }
        }
        #endregion

        [SetUp]
        public void SetUp()
        {
            _monitor = new VMMonitor();
        }

        [Test]
        public void DefaultReportIsFormatted()
        {
            //Act
            _monitor.Monitor(new MainVM(), "main");

            //Assert
            Approvals.Verify(_monitor.Report);
        }

        [Test]
        public void CollectionMemberUpdatesAreReported()
        {
            //Arrange
            var main = new MainVM();
            _monitor.Monitor(main, "main");

            //Act
            var x = 0;
            foreach (var member in main.Members)
            {
                member.Count = x++;
            }
            _monitor.ReportState(main);

            //Assert
            Approvals.Verify(_monitor.Report);
        }

        [Test]
        public void CollectionMemberUpdatesReportCorrectIndex()
        {
            //Arrange
            _monitor.WriteLine("First this test will remove items and update the last item in the Members collection.");
            _monitor.WriteLine("Second, it will replace the removed items and again, update the last item in the Members collection.");
            _monitor.WriteLine("Each last item update will set the item's Count to the index of the last item. This should match the reported index in all cases.");

            var main = new MainVM();
            _monitor.Monitor(main, "main");

            //Act
            var removed = new List<CollectionMember>();
            for (var x = 0; x < 5; ++x)
            {
                var member = main.Members[main.Members.Count - 1];
                member.Count = main.Members.Count - 1;

                var removeThisTime = main.Members[0];
                removed.Add(removeThisTime);
                main.Members.Remove(removeThisTime);
            }

            _monitor.ReportState(main);

            foreach (var restore in removed)
            {
                main.Members.Insert(0, restore);
                var member = main.Members[main.Members.Count - 1];
                member.Count = main.Members.Count - 1;
            }

            _monitor.ReportState(main);

            //Assert
            Approvals.Verify(_monitor.Report);
        }
    }

    internal class AcceptanceTestVM : INotifyPropertyChanged
    {
        public event PropertyChangedEventHandler PropertyChanged;

        [NotifyPropertyChangedInvocator]
        protected virtual void OnPropertyChanged(string propertyName)
        {
            var handler = PropertyChanged;
            if (handler != null) handler(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}
