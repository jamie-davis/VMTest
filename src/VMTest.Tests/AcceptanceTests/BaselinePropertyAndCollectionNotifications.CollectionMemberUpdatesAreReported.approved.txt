Accepted view model "main":                                                                    
                                                                                               
MainVM                                                                                         
------                                                                                         
                                                                                               
                                                                                               
    Members                                                                                    
    -------                                                                                    
    Name           Count                                                                       
    -------- -----------                                                                       
    Alpha              5                                                                       
    Beta               4                                                                       
    Charlie            7                                                                       
    Delta              5                                                                       
    Echo               4                                                                       
    Hotel              5                                                                       
    Indigo             6                                                                       
    Juliette           8                                                                       
    Kilo               4                                                                       
    Lima               4                                                                       
    Mike               4                                                                       
    November           8                                                                       
    Oscar              5                                                                       
    Papa               4                                                                       
    Quebec             6                                                                       
    Romeo              5                                                                       
    Sierra             6                                                                       
    Tango              5                                                                       
    Uniform            7                                                                       
    Victor             6                                                                       
    Whiskey            7                                                                       
    X-Ray              5                                                                       
    Zulu               4                                                                       
                                                                                               
                                                                                               
-->main.Members[0].Count = 0                                                                   
-->main.Members[1].Count = 1                                                                   
-->main.Members[2].Count = 2                                                                   
-->main.Members[3].Count = 3                                                                   
-->main.Members[7].Count = 7                                                                   
-->main.Members[8].Count = 8                                                                   
-->main.Members[9].Count = 9                                                                   
-->main.Members[10].Count = 10                                                                 
-->main.Members[11].Count = 11                                                                 
-->main.Members[12].Count = 12                                                                 
-->main.Members[13].Count = 13                                                                 
-->main.Members[14].Count = 14                                                                 
-->main.Members[15].Count = 15                                                                 
-->main.Members[16].Count = 16                                                                 
-->main.Members[17].Count = 17                                                                 
-->main.Members[18].Count = 18                                                                 
-->main.Members[19].Count = 19                                                                 
-->main.Members[20].Count = 20                                                                 
-->main.Members[21].Count = 21                                                                 
-->main.Members[22].Count = 22                                                                 
                                                                                               
VM "main" state:                                                                               
                                                                                               
MainVM                                                                                         
------                                                                                         
                                                                                               
                                                                                               
    Members                                                                                    
    -------                                                                                    
    Name           Count                                                                       
    -------- -----------                                                                       
    Alpha              0                                                                       
    Beta               1                                                                       
    Charlie            2                                                                       
    Delta              3                                                                       
    Echo               4                                                                       
    Hotel              5                                                                       
    Indigo             6                                                                       
    Juliette           7                                                                       
    Kilo               8                                                                       
    Lima               9                                                                       
    Mike              10                                                                       
    November          11                                                                       
    Oscar             12                                                                       
    Papa              13                                                                       
    Quebec            14                                                                       
    Romeo             15                                                                       
    Sierra            16                                                                       
    Tango             17                                                                       
    Uniform           18                                                                       
    Victor            19                                                                       
    Whiskey           20                                                                       
    X-Ray             21                                                                       
    Zulu              22                                                                       
                                                                                               
                                                                                               
                                                                                               